### ISSUE TITLE
Verus panic when revealing state machine transition definition: "internal error: generated ill-typed AIR code: error 'error 'error 'use of undeclared variable"

### ISSUE THREAD

**Issue by tenzinhl** [verus-lang/verus#411]
Verus panicked with strange output when I revealed the `next` function auto-generated by the `state_machine!` macro inside of a lemma.

**Verus version**: built using commit hash e2f56d1b26629ecc09c3ae5d19ce817da2ea277f (on main branch, no extra build flags, so debug build)

### How to reproduce

Repo: https://github.com/vmware-labs/verified-betrfs/tree/tenzinhl-reveal-panic
tag: `tenzinhl-reveal-panic`

(Link links to tag)

1. Clone the repo
2. cd `Splinter/src`
3. `${verus_command} bundle.rs`

Produces a panicked output, here's a snippet of the beginning:

"thread 'rustc' panicked at 'internal error: generated ill-typed AIR code: error 'error 'error 'use of undeclared variable fuel%coordination_layer.AbstractJournal_v.AbstractJournal.State.next.' in expression 'fuel%coordination_layer.AbstractJournal_v.AbstractJournal.State.next.'' in expression '(fuel_bool fuel%coordination_layer.AbstractJournal_v.AbstractJournal.State.next.)'' in statement '...
[... an expanded what looks like z3 statement]
"

Panics because of line 432 in `Splinter/src/coordination_layer/CoordinationSystemRefinement_v.rs` which reveals the `next` and `next_by` predicates auto-generated for a state machine made using the `state_machine!` macro.

### Full output

Full error message is in this file.
[reveal_panic_verus_issue.txt](https://github.com/verus-lang/verus/files/10864411/reveal_panic_verus_issue.txt)

**Comment by tenzinhl** [comment]
Found a more simplified reproduction. Seems like it's the specific `next` function that I'm trying to reveal that's causing problems for some reason (specifically `AbstractJournal::State::next`). Added a dummy function which calls the reveal, panic output is then much simpler. Code is tagged as `tenzinhl-reveal-panic-simpler` in the verified-betrfs repo.

Output here:
```
[tenzinhl@LAPTOP-U355A8BD:src]$ $verus --verify-module coordination_layer::CoordinationSystemRefinement_v bundle.rs
note: verifying module coordination_layer::CoordinationSystemRefinement_v

thread 'rustc' panicked at 'internal error: generated ill-typed AIR code: error 'error 'use of undeclared variable fuel%coordination_layer.AbstractJournal_v.AbstractJournal.State.next.' in expression 'fuel%coordination_layer.AbstractJournal_v.AbstractJournal.State.next.'' in expression '(fuel_bool fuel%coordination_layer.AbstractJournal_v.AbstractJournal.State.next.)'', rust_verify/src/verifier.rs:362:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
verification results:: verified: 1 errors: 0
```

**Comment by tjhance** [comment]
Here's a minimal example that makes Verus panic. Verus doesn't crash if you un-comment the `assert`, so I think the crash occurs because Verus doesn't include `AbstractJournal::State::next` in the module's transitive dependencies.

```rust
mod X {
    #[verifier(opaque)]
    pub open spec fn foo() -> bool { true }
}

proof fn test() {
    reveal(X::foo);
    //assert(X::foo());  // uncomment this line and it doesn't crash
    assert(false);
}
```

**Comment by tjhance** [comment]
Here's another tangential issue, it looks like we don't do any sanity-checking on `reveal`, for example, Verus crashes if you try to reveal an exec function:

```rust
pub fn some_exec_fn() -> bool { true }

proof fn test() {
    reveal(some_exec_fn);
}
```

It's the same AIR error about an undeclared variable, though the fix would be different in this case.

**Comment by tenzinhl** [comment]
Yep thanks for the help. Turns out a `require` clause in a transition was calling `next_by` on a composed state machine (whereas normally we were calling `next` (cleaner interface)). So yeah I guess no transitive dependency on `next` in that case. Switched the old definition of the transition to use `next` and now we're well on our way.

**Comment by tjhance** [comment]
fixed by #419
